File: insultbot.py
# Chance Snow
# Mar 31, 2013
# 2:00 AM
# irc handle: enigma
#######################
# An IRC bot that insults people
# Fetches insults from insultgenerator.org and randominsults.net
# handle: insultbot

import sys
import random
import urllib
import datetime
from os.path import expanduser
from HTMLParser import HTMLParser
from oyoyo.client import IRCClient
from oyoyo.cmdhandler import DefaultCommandHandler
from oyoyo import helpers

HOST         = 'iss.cat.pdx.edu'
PORT         = 6697
NICK         = 'insultbot'
REAL_NAME    = ':Enigma\'s Insult Bot'
CHANNEL      = '#insultbot'

SERVICES     = ['http://www.randominsults.net/']

class Logger(object):
    def __init__(self, filename='insultbot.log', console=True):
        self.terminal = sys.stdout
        self.filename = filename
        self.console = console

    def write(self, message):
        self.log = open(self.filename, 'a')
        if self.console:
            self.terminal.write(message)
        self.log.write(message)
        self.log.close()

#print only to log file
sys.stdout = Logger(expanduser('~/.bots/insultbot.log'), False)

#read channels from config file
CHANNELS = []
cfgFile = open(expanduser('~/.bots/.insultbot'), 'r')
for line in cfgFile:
    line = line.strip()
    if not line.startswith('--'):
        if line.find(' ') != -1:
            line = line.split(' ')
            CHANNELS.append((line[0], line[1]))
        else:
            CHANNELS.append((line, None))

print '================================================'
print 'Connecting ' + NICK + ' on ' + HOST + ':' + str(PORT)
now = datetime.datetime.now()
print now.strftime("%a, %b %d, %Y - %I:%M:%S %p")
print '------------------------------------------------'

class InsultHTMLParser(HTMLParser):
    def __init__(self, service):
        HTMLParser.__init__(self)
        self.service = service
        self.tag = None
        self.attrs = None
        self.insult = None
        self.found = False

    def handle_starttag(self, tag, attrs):
        self.tag = str(tag)
        self.attrs = attrs
    def hendle_endtag(self, tag): return
    def handle_data(self, data):
        #print 'Data in ' + str(self.tag) + ' element: ' + str(data)
        if self.service == SERVICES[0]:
            if str(self.tag) == 'i' and not self.found:
                self.insult = str(data).strip()
                self.found = True


def connect_callback(client):
    #helpers.msg(client, "NickServ", "IDENTIFY", "password")
    
    for chan in CHANNELS:
        channel = chan[0]
        key = chan[1]
        if key is None:
            helpers.join(client, channel)
        else:
            helpers.join(client, channel, key)
        print 'Joined ' + channel

    print '------------------------------------------------'


class EnigmaBotHandler(DefaultCommandHandler):
    def say(self, msg, to=None):
        if to == None: to = self.channel
        helpers.msg(self.client, to, msg)
   
    def getnick(self, nick):
        bangIndex = nick.find('!')
        if bangIndex != -1:
            return nick[:bangIndex]
        else:
            return nick

    def getinsult(self):
        #From a random service
        service = random.choice(SERVICES)
        response = urllib.urlopen(service)
        html = response.read()
        parser = InsultHTMLParser(service)
        parser.feed(html)
        insult = parser.insult
        if insult:
            return insult
        else:
            return random.choice(['You suck!', 'Screw you!'])

    def insult(self, to, private=True):
        print 'Insulting ' + to
        msg = 'Hey ' + to + '! ' + self.getinsult()
        if private:
            self.say(msg, to)
        else:
            self.say(msg)

    def privmsg(self, nick, channel, msg):
        self.channel = channel
        nick = self.getnick(nick)
        msg = msg.decode()

        if nick == NICK:
            return
        
        #leave Bot message
        if msg.find('!' + NICK + ' leave') != -1:
            for chan in CHANNELS:
                if chan[0] == channel:
                    self.insult(nick, False)
                    print "Leaving " + channel
                    helpers.part(self.client, channel)
                    break
        
        #quit Bot message
        if msg.find('!' + NICK + ' quit') != -1:
            for chan in CHANNELS:
                if chan[0] == channel:
                    self.insult(nick, False)
                    print "Leaving " + channel
                    helpers.part(self.client, channel)
                    print "Quitting IRC"
                    helpers.quit(self.client, 'Quitting at request')
                    quit()
        
        #help Bot message
        if msg.find('!' + NICK + ' help') != -1:
            for chan in CHANNELS:
                if chan[0] == channel:
                    msg = 'Insultbot Help\n====================\nSay anything about insultbot and prepare to be insulted!\n \nDISCLAIMER: Some of the isults may be graphic in nature.\nBy using insultbot you are agreeing to the following terms:\n  The creator of insultbot (Chance Snow <enigma>) shall not be held liable for anything insultbot says to anyone.\n \nInsultbot Commands:\n----------\n\'!insultbot insult <nick>\'  - Insult someone privately\n\'insult: <nick>\'             - Insult someone privately\n\'!insultbot greet <nick>\'   - Insult someone publicly'
                    self.say(msg, nick)
                    return

        if msg.lower().find('hello') != -1 and msg.lower().find(NICK) != -1:
            self.insult(nick, False)
            return
        
        if msg.lower().find('hi') != -1 and msg.lower().find(NICK) != -1:
            self.insult(nick, False)
            return
        
        if msg.find('!' + NICK + ' greet ') != -1:
            to = msg.split('greet ')[1]
            if to.find(NICK) == -1:
                self.insult(to, False)
            else:
                self.insult(nick, False)
            return
        
        if msg.find('!' + NICK + ' insult ') != -1:
            to = msg.split('insult ')[1]
            if to.find(NICK) != -1:
                self.insult(to)
            else:
                self.insult(nick)
            return
        
        if msg.find('insult: ') != -1:
            to = msg.split(' ')[1]
            if to.find(NICK) != -1:
                self.insult(to)
            else:
                self.insult(nick)
            return

        if msg.find(NICK) != -1 and msg.find('!' + NICK) == -1:
            print 'Insulting ' + nick
            self.say('[Err] ' + random.choice(
                ['I\'m sorry, did you say something <NICK>?',
                'What did you call me <NICK>?',
                'Leave me alone <NICK>!',
                'Did your fatass cat fall on your keyboard <NICK>?']).replace('<NICK>', nick))
    
    def kick(self, channel, nick, reason):
        for chan in CHANNELS:
            if chan[0] == channel:
                print 'Rejoining ' + channel
                if chan[1] is None:
                    helpers.join(self.client, channel)
                else:
                    helpers.join(self.client, channel, chan[1])
                break


#Create client connection to The CAT's IRC server
client = IRCClient(EnigmaBotHandler,
        host=HOST, port=PORT, nick=NICK, real_name=REAL_NAME, blocking=True, ssl=True,
        connect_cb=connect_callback)

connection = client.connect()
while True:
    connection.next()

File: insultbot.rb
# InsultBot
########################################################
# Written by Chance Snow - enigma <chances@cat.pdx.edu>
# Licensed under the MIT License
# See LICENSE file for more info
########################################################

require 'cinch'
require 'cinch/plugins/identify'
require 'yaml'
require 'date'
require 'open-uri'
require 'nokogiri'

@debug = false
if ARGV.length > 0
  if ARGV.include?('--debug')
    @debug = true
    puts 'DEBUG MODE'
  end
end

@channels = nil
@debug_channels = nil

insultbot = Cinch::Bot.new do
  configure do |c|
    #Load configuration
    config = {}
    if File.exists?(File.expand_path('~/.bots/insultbot.yml'))
      config = YAML::load_file(File.expand_path('~/.bots/insultbot.yml'))
    end
    
    if not config.has_key?('channels')
      @channels = ['#insultbot']
    else
      @channels = config['channels']
      if not @channels.is_a?(Array) or config['channels'].length == 0
        @channels = ['#insultbot']
      end
    end
    if not config.has_key?('debug') or not config['debug'].has_key?['channels']
      @channels = ['#insultbot']
    else
      @debug_channels = config['debug']['channels']
      if not @debug_channels.is_a?(Array) or config['debug']['channels'].length == 0
        @debug_channels = ['#insultbot']
      end
    end
    #Configure bot
    c.server = 'iss.cat.pdx.edu'
    c.port = 6697
    c.ssl.use = true
    c.nick = 'insultbot'
    c.realname = 'Enigma\'s InsultBot'
    if ARGV.include?('--debug')
      c.channels = @debug_channels
    else
      c.channels = @channels
    end
    c.messages_per_second = 20
    if config.has_key?('nickserv') and config['nickserv'].has_key?['password']
      c.plugins.plugins = [Cinch::Plugins::Identify]
      c.plugins.options[Cinch::Plugins::Identify] = {
        :password => config['nickserv']['password'],
        :type => :nickserv
      }
    end
  end
  
  helpers do
    @help = 'See http://web.cecs.pdx.edu/~chances/insults/help'
    @thanks = ['Thanks!', 'Much obliged.', 'Cheers!', 'Thank you.', 'Cool karma.']
    @minus = ['What gives?', 'That was uncalled for!', 'Not cool.', 'Was it something I said?']
    @services = [
      'http://www.randominsults.net/',
      'http://programmerinsults.com/',
      'http://web.cecs.pdx.edu/~chances/insults/insult/random?format=text'
    ]
    
    def help(user)
      for line in @help.split('\n')
        user.msg line
      end
    end
    
    def insult(to)
      #Get an insult
      service = @services.sample
      insult = nil
      case service
      when @services[0] #Random Insults.net
        page = Nokogiri::HTML(open(service))
        insult = page.css('i').first.text
      when @services[1] #Programmer Insults.com
        page = Nokogiri::HTML(open(service))
        insult = page.css('a').first.text
      when @services[2] #InsultBot Suggestions
        open(service) do |file|
          insult = file.read
        end
      else
        page = Nokogiri::HTML(open(@services[0]))
        insult = page.css('i').first.text
      end
      
      #Insult the user
      if not insult == nil
        info "Insulting #{to}"
        if insult.include? '<NICK>'
          insult.gsub('<NICK>', to)
        else
          "Hey #{to}! #{insult}"
        end
      else
        info "ERROR: Failed to retrieve an insult"
      end
    end

    def handle_command(m, command, args)
      from = m.user.nick
      case command
      when /help/
        m.reply(@help)
      when /(greet|insult)/
        if not args.include?(m.bot.nick) and args.length > 0
          if args == 'me'
            m.reply insult(from)
          else
            m.reply insult(args)
          end
        else
          m.reply insult(from)
        end
      when /leave/
        if from == 'enigma'
          m.reply insult(from)
          info "Leaving #{m.channel}"
          m.bot.part(m.channel)
        end
      when /quit/
        if from == 'enigma'
          m.reply insult(from)
          info 'Quitting IRC'
          m.bot.quit('Quitting at the request of my master')
        end
      else
        m.reply "[Error] #{insult(from)}"
      end
    end
  end
  
  on :connect do |m|
    info '------------------------------------------------'
    for channel in m.bot.config.channels
      info "Joining #{channel.split(' ')[0]}"
    end
    info '------------------------------------------------'
  end
  
  on :channel, /^!help insultbot$/ do |m|
    m.reply(@help)
  end
  on :channel, /^!insultbot (\w+)\s*(.*)/ do |m, command, args|
   handle_command(m, command, args)
  end
  on :channel, /^insultbot: (\w+)\s*(.*)/ do |m, command, args|
   handle_command(m, command, args)
  end
  on :private, /^(\w+)\s*(.*)/ do |m, command, args|
   handle_command(m, command, args)
  end
  on :message, /^insult: (.+)/ do |m, who|
    if who == 'me'
      m.reply insult(m.user.nick)
    elsif not who.include?(m.bot.nick)
      m.reply insult(who)
    else
      m.reply insult(m.user.nick)
    end
  end
  on :message, /(insultbot|InsultBot)/ do |m|
    from = m.user.nick
    return if from == m.bot.nick
    unless m.message.match(/^(!insultbot.*|insultbot: .*|!help insultbot|insult: .*|infoobot: .*)/)
      if m.message.match(/(insultbot|InsultBot)\+\+/)
        m.reply "#{from}: #{@thanks.sample}"
      elsif m.message.match(/(insultbot|InsultBot)--/)
        m.reply "#{from}: #{@minus.sample}"
      else
        #TODO: Make some insults actions.
        if m.action?
          m.reply insult(from)
        else
          m.reply insult(from)
        end
      end
    end
  end
end

#Set log file
log_file = File.open(File.expand_path('~/.bots/insultbot.log'), 'a')
log_file.sync = true
insultbot.loggers << Cinch::Logger.new(log_file)
insultbot.loggers.level = :info
insultbot.loggers.first.level  = :info

log_file.puts '================================================'
log_file.puts DateTime.now.strftime('%a, %b %d, %Y - %I:%M:%S %p')
log_file.puts '------------------------------------------------'

if @debug
  log_file.puts 'DEBUG MODE'
  log_file.puts '------------------------------------------------'
end

insultbot.start
